sdk/Pulumi.Tests/Provider/PropertyValueTests.cs

```diff
@@ -0,0 +1,310 @@
+        var serializer = new PropertyValueSerializer();
+        var secretValue = new PropertyValue(new PropertyValue("Hello"));
+
+        var deserialized = await serializer.Deserialize<Input<string>>(secretValue);
+        var output = deserialized.ToOutput();
+        var data = await output.DataTask;
+        Assert.Equal("Hello", data.Value);
+        Assert.True(data.IsSecret);
+        Assert.True(data.IsKnown);
+    }
+
+    [Fact]
+    public async Task DeserializingBasicInputsWorks()
+    {
+        var serializer = new PropertyValueSerializer();
+        var resources = ImmutableArray.Create("pulumi:pulumi:Stack");
+
+        var output = new PropertyValue(new OutputReference(
+            value: new PropertyValue("Hello"),
+            dependencies: resources));
+
+        var deserialized = await serializer.Deserialize<Input<string>>(output);
+        var deserializedOutput = deserialized.ToOutput();
+        var data = await deserializedOutput.DataTask;
+        Assert.Equal("Hello", data.Value);
+        Assert.False(data.IsSecret);
+        Assert.True(data.IsKnown);
+        Assert.Single(data.Resources);
+        if (data.Resources.First() is DependencyResource dependencyResource)
+        {
+            var resolvedUrn = await dependencyResource.Urn.GetValueAsync("");
+            Assert.Equal("pulumi:pulumi:Stack", resolvedUrn);
+        }
+        else
+        {
+            throw new Exception("Expected deserialized resource to be a dependency resource");
+        }
+    }
+
+    [Fact]
+    public async Task DeserializingWrappedOutputSecretWorks()
+    {
+        var serializer = new PropertyValueSerializer();
+        // secretOutput = Secret(Output("Hello"))
+        var secretOutput = new PropertyValue(
+            new PropertyValue(new OutputReference(
+                value: new PropertyValue("Hello"),
+                dependencies: ImmutableArray<string>.Empty)));
+
+        var deserialized = await serializer.Deserialize<Input<string>>(secretOutput);
+        var deserializedOutput = deserialized.ToOutput();
+        var data = await deserializedOutput.DataTask;
+        Assert.Equal("Hello", data.Value);
+        Assert.True(data.IsSecret);
+        Assert.True(data.IsKnown);
+    }
+
+    [Fact]
+    public async Task DeserializingWrappedSecretInOutputWorks()
+    {
+        var serializer = new PropertyValueSerializer();
+        // secretOutput = Output(Secret("Hello"))
+        var secretOutput = new PropertyValue(
+            new OutputReference(
+                value: new PropertyValue(new PropertyValue("Hello")),
+                dependencies: ImmutableArray<string>.Empty));
+
+        var deserialized = await serializer.Deserialize<Input<string>>(secretOutput);
+        var deserializedOutput = deserialized.ToOutput();
+        var data = await deserializedOutput.DataTask;
+        Assert.Equal("Hello", data.Value);
+        Assert.True(data.IsSecret);
+        Assert.True(data.IsKnown);
+    }
+}
```

Comment:
We're missing tests of the Input/Output attributes to rename fields.

Reply:
Added those ✅